### PR TITLE
Maven CLI: add startWith matching into recipes detection

### LINKED ISSUE #36573: Maven CLI: add wildcard matching into recipes detection
### Describe the bug

The recent change #34391, bringing the detection of extensions based on dependencies, revealed a problem with transitive dependencies.

Custom recipies for the Camel-Quarkus are executed only if `camel-quarkus-core` is present in the pom as a top level dependency.

In the reality, the camel-quarkus project contains usually dependencies for the extensions (like `camel-quarkus-file`), which bring `camel-quarkus-core` transitively.
Such project is not updated correctly, because CQ-core dependency is not detected and CQ recipes are not applied.

What do I suggest?

The dependency groupId/artifactId is matched with the folder structure of recipes (see https://github.com/quarkusio/quarkus-updates/tree/main/recipes/src/main/resources/quarkus-updates/org.apache.camel.quarkus/camel-quarkus-core)
Simple solution would be to use a "wildcard" in the name of the artifactId folder. For example `org.apache.camel.quarkus/camel-quarkus-*`. Such folder would be matched both by `camel-quarkus-file` or `camel-quarkus-core`

(We could use also `startsWith` instead of `equals` and keep the name of the folder `camel-quarkus-`, but this approach might not be desired. (for example `ext-custom` and `ext-custom-better` should not be matched together)

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by JiriOndrusek**
fixes https://github.com/quarkusio/quarkus/issues/36573

Description:

1 - Implementation support only wildcard ('*') at the end of the artifactId name (directory representing artifactId name)
2 - Change is back-compatible with the recipes without wildcards ( which is probably common usage for extension recipes)
3 - Test covering several cases are present.
4 - I changed internal implementation from `List<String>` to Map<String, String> to avoid duplication of recipes loaded because of directories with wildcards. (I kept amending of the content in case it differs - can not happen in this implementation, but for the future changes it might be helpful)  
5 - If the quarkus CLI is used with the older recipes (without wildcards) - everything works as expected.
     If the Quakus CLI prior this change is used with the recipes using wildcards, recipes won't be matched, but no internal error will happen. I think that it is the desired behavior.

### CODE REVIEW

**Comment by JiriOndrusek** [comment]
FYI @ia3andy

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/update/rewrite/QuarkusUpdatesRepository.java [hunk 1]
```diff
@@ -56,7 +58,7 @@ public static FetchResult fetchRecipes(MessageWriter log, MavenArtifactResolver
             final Artifact artifact = artifactResolver.resolve(DependencyUtils.toArtifact(gav)).getArtifact();
             final ResourceLoader resourceLoader = ResourceLoaders.resolveFileResourceLoader(
                     artifact.getFile());
-            final List<String> recipes = fetchRecipesAsList(resourceLoader, "quarkus-updates", recipeDirectoryNames);
+            final Map<String, String> recipes = fetchRecipesAsList(resourceLoader, "quarkus-updates", recipeDirectoryNames);
```
Comment (ia3andy) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/update/rewrite/QuarkusUpdatesRepository.java:61]: just call it `fetchUpdateRecipes`

**Review by ia3andy — CHANGES_REQUESTED** [review]
Thanks, just a few comments

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/update/rewrite/QuarkusUpdatesRepository.java (hunk 1 above)
Comment (JiriOndrusek) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/update/rewrite/QuarkusUpdatesRepository.java]: fixed

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/update/rewrite/QuarkusUpdatesRepository.java [hunk 2]
```diff
@@ -133,38 +135,69 @@ static boolean shouldApplyRecipe(String recipeFileName, String currentVersion, S
...
-                                                            return new String(Files.readAllBytes(p));
-                                                        } catch (IOException e) {
-                                                            throw new RuntimeException("Error reading file: " + p, e);
-                                                        }
-                                                    })
-                                                    .onClose(() -> recipePath.close());
-                                        } catch (IOException e) {
-                                            throw new RuntimeException("Error traversing directory: " + dir, e);
-                                        }
+                                    String regexKey = toKey(path.relativize(dir).toString());
```
[the remark this answers is not in the document]
Comment (JiriOndrusek) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/update/rewrite/QuarkusUpdatesRepository.java]: I extracted `applyWildcard` and `toKey` is already there, so it is possible to test. `applyWildcard` is already tested in `QuarkusUpdatesRepositoryTest.java`, for `toKey` there is an empty method (I have to find a parameters relevant for testing yet)

[the remark this answers is not in the document]
Comment (JiriOndrusek) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/update/rewrite/QuarkusUpdatesRepository.java]: I wrote test content in `testToKey` method. 

Problem fixed.

**Review by ia3andy — APPROVED** [review]
cool!

**Comment by ia3andy** [comment]
Ok, the problem is with `*` in directories path which is not ok on Windows. The idea to fix this is to change Wildcard by a startWith that we always use.

**Comment by ia3andy** [comment]
@JiriOndrusek you will need to also update the PR on quarkus-updates

**Comment by JiriOndrusek** [comment]
Thanks for the investigation. I'll refactor the solution to use `startsWith` and the change the related PR for Quarkus-updates

**Comment by JiriOndrusek** [comment]
@ia3andy PR is refactored to use startsWith instead of ..*, PR in quarkus-updates is changed accordingly.

**Review by ia3andy — APPROVED** [review]
Thanks for the quick change @JiriOndrusek

**Comment by JiriOndrusek** [comment]
@ia3andy I checked the failed CI build and I can still see `*` in the log, eventhough the cirrent change does not contain such folder anymore
`java.nio.file.InvalidPathException: Illegal char <*> at index 80: /home/app/target/classes/quarkus-updates/org.apache.camel.quarkus.camel-quarkus-*`

**Comment by JiriOndrusek** [comment]
Sorry, I see the reason, my fault. I'll fix it shorthly

**Comment by JiriOndrusek** [comment]
I'll also rebase the PR

**Comment by JiriOndrusek** [comment]
@ia3andy There was 1 failing test because of the Windows system. I fixed it to work on windows (therefore failing on linux) and enabled it only on windows os.

The problem was happening in `toKey`  

``` 
org.opentest4j.AssertionFailedError: expected: <..:C::a:b:quarkus-updates:org.apache.camel.quarkus.camel-quarkus> but was: <..:b:quarkus-updates:org.apache.camel.quarkus.camel-quarkus>
```

**Comment by JiriOndrusek** [comment]
Small typo in my last pushed was fixed, unfortunately the CI has to be approved again.

**Comment by JiriOndrusek** [comment]
Hi @ia3andy , from the current failure I don't see any relation to the code of this PR (But I'm not able to find a reason of failure for 2 runs so far), WDYT?

**Comment by JiriOndrusek** [comment]
I see:

JDK11
```
2023-10-30T16:45:44.8306526Z [INFO] Quarkus - Integration Tests - Hibernate ORM - Compatibility with databases meant for ORM 5.6 - MariaDB FAILURE [ 25.639 s]
2023-10-30T16:45:44.8308312Z [INFO] Quarkus - Integration Tests - Hibernate ORM - Compatibility with databases meant for ORM 5.6 - PostgreSQL FAILURE [ 19.750 s]
```

JDK17
```

```

JDK21
```

```

```
2023-10-30T16:30:29.0798092Z [INFO] Quarkus - Integration Tests - SmallRye Context Propagation FAILURE [ 34.797 s]
```

**Comment by JiriOndrusek** [comment]
From the previous comment, none of the failures are not related to the change

**Comment by ia3andy** [comment]
They don't seem to be related, I'll merge

**Pull request merged**